lib/web/integrations_awsoidc.go

```diff
@@ -372,6 +377,132 @@ func (h *Handler) awsOIDCListSecurityGroups(w http.ResponseWriter, r *http.Reque
+		return nil, trace.Wrap(err)
+	}
+
+	resp, err := awsOIDCRequiredVPCSHelper(ctx, req, listDBsClient, clt)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return resp, nil
+}
+
+func awsOIDCRequiredVPCSHelper(ctx context.Context, req ui.AWSOIDCRequiredVPCSRequest, listDBsClient awsoidc.ListDatabasesClient, clt auth.ClientI) (*ui.AWSOIDCRequiredVPCSResponse, error) {
+	resp, err := awsoidc.ListAllDatabases(ctx, listDBsClient, req.Region)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(resp.Databases) == 0 {
+		return nil, trace.BadParameter("there are no available RDS instances or clusters found in region %q", req.Region)
+	}
+
+	// Get all database services with ecs/fargate metadata label.
+	nextToken := ""
+	fetchedDbSvcs := []types.DatabaseService{}
+	for {
+		page, err := client.GetResourcePage[types.DatabaseService](ctx, clt, &proto.ListResourcesRequest{
+			ResourceType: types.KindDatabaseService,
+			Limit:        defaults.MaxIterationLimit,
+			StartKey:     nextToken,
+			Labels:       map[string]string{types.AWSOIDCAgentLabel: types.True},
+		})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		fetchedDbSvcs = append(fetchedDbSvcs, page.Resources...)
+		nextToken = page.NextKey
+		if len(nextToken) == 0 {
+			break
+		}
+	}
+
+	// Construct map of VPCs and its subnets.
+	vpcLookup := map[string][]string{}
+	for _, db := range resp.Databases {
+		rds := db.GetAWS().RDS
+		vpcId := rds.VPCID
+		if _, found := vpcLookup[vpcId]; !found {
+			vpcLookup[vpcId] = rds.Subnets
+			continue
+		}
+		combinedSubnets := append(vpcLookup[vpcId], rds.Subnets...)
+		vpcLookup[vpcId] = utils.Deduplicate(combinedSubnets)
+	}
+
+	// Start looking for db service matches.
+	wantLabels := map[string][]string{
+		types.DiscoveryLabelAccountID: {},
+		types.DiscoveryLabelRegion:    {},
+		types.DiscoveryLabelVPCID:     {},
+	}
+
+	for _, svc := range fetchedDbSvcs {
+		// Create a lookup table of labels for easier matching.
+		labelLookup := map[string][]string{}
+		for _, matcher := range svc.GetResourceMatchers() {
+			for key, newVals := range *matcher.Labels {
+				if existingVals, ok := labelLookup[key]; ok {
+					labelLookup[key] = append(existingVals, newVals...)
+					continue
+				}
+				labelLookup[key] = newVals
+			}
+		}
+
+		// Do an exact match, b/c other labels may not match.
+		if len(labelLookup) != len(wantLabels) {
+			continue
+		}
+
+		// Match labels contains the keys we are looking for.
+		matchedLabelKeys := true
+		for key := range wantLabels {
+			vals, found := labelLookup[key]
+			if !found {
+				matchedLabelKeys = false
+				break
+			}
+			wantLabels[key] = vals
+		}
+
+		if matchedLabelKeys &&
+			slices.Contains(wantLabels[types.DiscoveryLabelAccountID], req.AccountID) &&
+			slices.Contains(wantLabels[types.DiscoveryLabelRegion], req.Region) {
+			// Delete found vpcs
+			vpcs := wantLabels[types.DiscoveryLabelVPCID]
+			for _, vpc := range vpcs {
+				delete(vpcLookup, vpc)
+			}
+		}
+	}
```

Comment:
I would do some assumptions here: we are only interested in the DatabaseServices deployed by us, which have a known configuration.

We should only have a single ResourceMatcher with non-empty Label set, and each Label must have a single LabelValue.
This should allow us to simplify the multiple for loops we have right now.

```suggestion
	for _, svc := range fetchedDbSvcs {
		if len(svc.GetResourceMatchers()) != 1 || svc.GetResourceMatchers()[0].Labels == nil {
			continue
		}
		labelMatcher := *svc.GetResourceMatchers()[0].Labels
		if len(labelMatcher) != 3 {
			continue
		}
		if slices.Compare(labelMatcher[types.DiscoveryLabelAccountID], []string{req.AccountID}) != 0 {
			continue
		}
		if slices.Compare(labelMatcher[types.DiscoveryLabelRegion], []string{req.Region}) != 0 {
			continue
		}
		if len(labelMatcher[types.DiscoveryLabelVPCID]) != 1 {
			continue
		}
		delete(vpcLookup, labelMatcher[types.DiscoveryLabelVPCID][0])
	}
```

Tests are failing with this change, but that's because we have multiple ResourceMatchers on the DatabaseServices configuration we created